js/libs/form.js

```diff
@@ -1455,58 +1451,72 @@ Fliplet().then(function() {
                   }
                 },
                 options: function(values) {
-                  if (typeof values === 'undefined') {
-                    return field.options;
-                  }
+                  var result = typeof values === 'function' ? values() : values;
 
-                  if (!Array.isArray(values)) {
-                    throw new Error('Options must be an array');
+                  if (typeof result === 'undefined') {
+                    return field.options;
                   }
 
-                  if (field._type === 'flSelect') {
-                    // remove all invalid options
-                    _.remove(values, function(val) {
-                      return !(_.isObject(val) || _.isNumber(val) || (_.isString(val) && val.trim()));
-                    });
-                  }
+                  return Promise.resolve(result).then(function(newOptions) {
+                    if (!Array.isArray(newOptions)) {
+                      throw new Error('Options must be an array');
+                    }
 
-                  var options = values.map(function(option) {
-                    if (typeof option === 'object') {
-                      if (typeof option.value !== 'undefined') {
-                        option.id = option.value;
-                      }
+                    values = newOptions || [];
 
-                      return option;
+                    if (field._type === 'flSelect') {
+                      // remove all invalid options
+                      _.remove(values, function(val) {
+                        return !(_.isObject(val) || _.isNumber(val) || (_.isString(val) && val.trim()));
+                      });
                     }
 
-                    return { id: option };
-                  });
+                    var options = values.map(function(option) {
+                      if (typeof option === 'object') {
+                        if (typeof option.value !== 'undefined') {
+                          option.id = option.value;
+                        }
 
-                  if (!_.isEmpty(field.value)) {
-                    switch (field._type) {
-                      case 'flCheckbox':
-                        var selectedValues = _.difference(field.value, values);
+                        if (field._type === 'flTypeahead' && typeof option.id === 'undefined') {
```

Comment:
What are you trying to do with all these logic that's treating the typeahead field differently?

Reply:
@tonytlwu I was having issue from selectize.js. It's required to have `id` in the options

Reply:
I see. Yes I think it does need to have IDs. You can use the same ID as the label (unless the user has provided a custom ID).